// angular-ng-autocomplete/esm5/lib/autocomplete-lib.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/autocomplete-lib.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { AutocompleteLibComponent } from './autocomplete-lib.component';
import { AutocompleteComponent } from './autocomplete/autocomplete.component';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { HighlightPipe } from './autocomplete/highlight.pipe';
import * as ɵngcc0 from '@angular/core';
var AutocompleteLibModule = /** @class */ (function () {
    function AutocompleteLibModule() {
    }
AutocompleteLibModule.ngModuleDef = ɵngcc0.ɵɵdefineNgModule({ type: AutocompleteLibModule });
/*@__PURE__*/ ɵngcc0.ɵɵsetNgModuleScope(AutocompleteLibModule, { declarations: function () { return [AutocompleteLibComponent, AutocompleteComponent, HighlightPipe]; }, imports: function () { return [CommonModule,
        FormsModule]; }, exports: function () { return [AutocompleteLibComponent, AutocompleteComponent, HighlightPipe]; } });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(AutocompleteLibModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule
                ],
                declarations: [AutocompleteLibComponent, AutocompleteComponent, HighlightPipe],
                exports: [AutocompleteLibComponent, AutocompleteComponent, HighlightPipe]
            }]
    }], function () { return []; }, null);
AutocompleteLibModule.ngInjectorDef = ɵngcc0.ɵɵdefineInjector({ factory: function AutocompleteLibModule_Factory(t) { return new (t || AutocompleteLibModule)(); }, imports: [[
            CommonModule,
            FormsModule
        ]] });
    return AutocompleteLibModule;
}());
export { AutocompleteLibModule };
