// angular-ng-autocomplete/esm5/lib/autocomplete-lib.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/autocomplete-lib.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var AutocompleteLibComponent = /** @class */ (function () {
    function AutocompleteLibComponent() {
    }
    /**
     * @return {?}
     */
    AutocompleteLibComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /** @nocollapse */
    AutocompleteLibComponent.ctorParameters = function () { return []; };
AutocompleteLibComponent.ngComponentDef = ɵngcc0.ɵɵdefineComponent({ type: AutocompleteLibComponent, selectors: [["ng-autocomplete-lib"]], factory: function AutocompleteLibComponent_Factory(t) { return new (t || AutocompleteLibComponent)(); }, consts: 2, vars: 0, template: function AutocompleteLibComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "p");
        ɵngcc0.ɵɵtext(1, " autocomplete-lib works! ");
        ɵngcc0.ɵɵelementEnd();
    } }, encapsulation: 2 });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(AutocompleteLibComponent, [{
        type: Component,
        args: [{
                selector: 'ng-autocomplete-lib',
                template: "\n    <p>\n      autocomplete-lib works!\n    </p>\n  "
            }]
    }], function () { return []; }, { ngOnInit: [] });
    return AutocompleteLibComponent;
}());
export { AutocompleteLibComponent };
